### ISSUE TITLE
False positive for switch expressions in MissingSwitchDefault

### ISSUE THREAD

**Issue by KENNYSOFT** [checkstyle/checkstyle#10866]
I have read check documentation: https://checkstyle.org/config_coding.html#MissingSwitchDefault

>Note that the compiler requires switch expressions to be exhaustive, so this check does not enforce default branches on such expressions.

```bash
/var/tmp $ javac Switch.java

/var/tmp $ cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="MissingSwitchDefault"/>
    </module>
</module>

/var/tmp $ cat Switch.java
public class Switch {
    public enum Options {
        ONE,
        TWO,
        THREE
    }

    public void foo1(Options option) {
        int num = switch (option) { // ok
            case ONE:
                yield 1;
            case TWO:
                yield 2;
            case THREE:
                yield 3;
        };
    }

    public void foo2(Options option) {
        assert Integer.valueOf(1).equals(switch (option) { // ok
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        });
    }

    public void bar1(Options option) {
        int num;
        num = switch (option) { // violation, but not expected 
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        };
    }

    public void bar2(Options option) {
        assert 1 < switch (option) { // violation, but not expected 
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        };
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.0.1-all.jar -c config.xml Switch.java
Starting audit...
[ERROR] /var/tmp/Switch.java:29:15: switch without "default" clause. [MissingSwitchDefault]
[ERROR] /var/tmp/Switch.java:37:20: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 2 errors.
```

<details>
<summary>Run script</summary>

```bash
cat > Switch.java <<EOF
public class Switch {
    public enum Options {
        ONE,
        TWO,
        THREE
    }

    public void foo1(Options option) {
        int num = switch (option) { // ok
            case ONE:
                yield 1;
            case TWO:
                yield 2;
            case THREE:
                yield 3;
        };
    }

    public void foo2(Options option) {
        assert Integer.valueOf(1).equals(switch (option) { // ok
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        });
    }

    public void bar1(Options option) {
        int num;
        num = switch (option) { // violation, but not expected 
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        };
    }

    public void bar2(Options option) {
        assert 1 < switch (option) { // violation, but not expected 
            case ONE -> 1;
            case TWO -> 2;
            case THREE -> 3;
        };
    }
}
EOF

javac Switch.java

cat > config.xml <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="MissingSwitchDefault"/>
    </module>
</module>
EOF

curl -O https://github.com/checkstyle/checkstyle/releases/download/checkstyle-9.0.1/checkstyle-9.0.1-all.jar
java -jar checkstyle-9.0.1-all.jar -c config.xml Switch.java
```

</details>

---

Expected: should not report line 29, 37 as violation.

![image](https://user-images.githubusercontent.com/6624567/137505248-4ae2cb09-7ed3-4447-b81a-d8076dfb0251.png)

In first false positive case, the parent type is 80(`TokenTypes.ASSIGN`) and it's different with `TokenTypes.EXPR`, `MissingSwitchDefaultCheck.isSwitchExpression()` returns `false`.

Reference: #8779

**Comment by nmancus1** [comment]
@KENNYSOFT I can see why this seems like a false positive, but I think that it is consistent with the check documentation and current Checkstyle limitations. From https://checkstyle.org/config_coding.html#MissingSwitchDefault : `...this way the code is protected against later changes`. Adding a default clause here *would* protect against future changes. In your example, `Options` is defined in the same file; however, Options could just as easily be defined elsewhere in another file. Herein lies the problem.

From https://checkstyle.org/writingchecks.html#Limitations : `You cannot see the content of other files.` Although the compiler would make sure that this switch expression is exhaustive, and thus technically makes the `default` clause redundant, without Checkstyle being able to see the content of other files, it is impossible for Checkstyle to know this.

**Comment by nmancus1** [comment]
In https://github.com/checkstyle/checkstyle/issues/8779 , I did not consider operators other than `=` as being a parent to a switch expression. Also we must have missed some nuance in regards to separate declaration and assignment.

**Comment by nmancus1** [comment]
> So the simple suggestion is following...

Looks like a reasonable place to start, feel free to send a PR once issue is approved.

**Comment by romani** [comment]
Approved.
Is it regression from 9.0 ?

**Comment by sukolenvo** [comment]
I am still getting this error with latest (9.2.1) build. Code to reproduce:
```java
public class Example {

  void run(Direction direction) {
    switch (direction) {
      case UP -> {
        System.out.println("Going up");
      }
      case DOWN -> {
        System.out.println("Going down");
      }
    }
  }

  enum Direction {
    UP, DOWN
  }
}
```

Config I am using:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
  <module name="TreeWalker">
    <module name="MissingSwitchDefault"/>
  </module>
</module>
```

Output:
```
Starting audit...
[ERROR] /home/gtx/Downloads/Example.java:4:5: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 1 errors.
```
Should I create a new issue or reopen this or #10463?

**Comment by romani** [comment]
@KENNYSOFT , please create new issue.

**Comment by romani** [comment]
@sukolenvo, please create new issue.

**Comment by KENNYSOFT** [comment]
@sukolenvo That's not a switch expression, but a switch statement. `->` is just for there are not any implicit continue after the label ends.

**Comment by sukolenvo** [comment]
Thank you work explaining, makes sense @KENNYSOFT
